sdks/python/apache_beam/io/gcp/gcsio.py

```diff
@@ -530,19 +541,63 @@ def _updated_to_seconds(updated):
         updated.microsecond / 1000000.0)
 
 
+def increment_bucket_read_counter_metric(read):
+  def inner(self, *args, **kwargs):
+    bytesRead = read(self, *args, **kwargs)
+    if self.enable_read_bucket_metric:
+      Metrics.counter(
+          self.__class__,
+          "GCS_read_bytes_counter_" + self._blob.bucket.name).inc(
+              len(bytesRead))
+    return bytesRead
+
+  return inner
+
+
 class BeamBlobReader(BlobReader):
-  def __init__(self, blob, chunk_size=DEFAULT_READ_BUFFER_SIZE):
+  def __init__(
+      self,
+      blob,
+      chunk_size=DEFAULT_READ_BUFFER_SIZE,
+      enable_read_bucket_metric=False):
     super().__init__(blob, chunk_size=chunk_size)
+    self.enable_read_bucket_metric = enable_read_bucket_metric
     self.mode = "r"
 
+  @increment_bucket_read_counter_metric
```

Comment:
Decorator here seems unnecessary. Already overwrite read() and write() method and the counter increment could just happen there

The refactor should also consider minimize the overhead when bucket_metric is not enabled.